sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -193,10 +193,12 @@ private[sql] object Dataset {
  */
 @Stable
 class Dataset[T] private[sql](
-    @DeveloperApi @Unstable @transient val queryExecution: QueryExecution,
+    @DeveloperApi @Unstable @transient private[sql] var queryExecution: QueryExecution,
```

Comment:
It's not thread-safe.

Reply:
This is marked as developer API, we can't make it private.

Reply:
I didn't find any description about accessibility in the comment of `@DeveloperApi`. Is this a tacit agreement among developers?